### ISSUE TITLE
site.master changing

### ISSUE THREAD

**Issue by mattaezell** [xcat2/xcat-core#4365]
It seems like every time we upgrade xCAT, our `site.master` gets changed to a different interface on our management node.  The interface is our bmc-in-os port, and it doesn't even have an entry in /etc/hosts or DNS.  It looks like various RPM pre/post install scripts call `xcatconfig`, and I suspect that's where this is happening.
- How can we prevent this from changing when upgrading xCAT?, or
- How do we make it pick the correct address?

**Comment by whowutwut** [comment]
@zet809 @robin2008 can you investigate this issue?

**Comment by robin2008** [comment]
It should be a bug, `xcatconfig` should always use the existing `site.master` first.

**Comment by robin2008** [comment]
In `xcatconfig`, the site table will only be populated with below conditions.
```
    # create basic site definition, if they do not exist reset request
    if (((!-r "/etc/xcat/site.sqlite") && (!-r "/etc/xcat/cfgloc"))
        || $::FORCE
        || $::initDB)
    {
```
The only things there might cause the issue might be `/etc/xcat/cfgloc` and `/etc/xcat/site.sqlite` is not there.

And I have tried that update site.master and run `xcatconfig -u` on an non-HA environment,  it looks the site.master will not be changed during the running of `xcatconfig -u`.

@mattaezell  I assume that you are running on an HA environment, right? 
What's your step to do the xCAT upgrade?  yum update on both MNs in the same time or 
update in on active MN, and then failover to the standby, and update it accordingly. (In HA, the standby Node might not mount /etc/xcat in standby mode)

But I thought this will only impact the local Sqlite DB, if using the PostgreSQL, the site.master should be still the former one.

**Comment by robin2008** [comment]
For the default site.master in `xcatconfig`, it will get the name by `hostname` command, and try to get its IP.  Could you check your MN to see the name resolution is what you expected?

**Comment by mattaezell** [comment]
Hello @robin2008
We have two management nodes on this system, but only one is active.  We use Postgres.

```
[root@mgmt1 db]# hostname
mgmt1.summit.olcf.ornl.gov
[root@mgmt1 db]# host mgmt1.summit.olcf.ornl.gov
mgmt1.summit.olcf.ornl.gov has address 10.10.0.4
[root@mgmt1 db]# host 10.10.0.4
4.0.10.10.in-addr.arpa domain name pointer mgmt1.summit.olcf.ornl.gov.
```

That is the correct IP (.4), and the one that should be in the database.

**Comment by robin2008** [comment]
@mattaezell 
It looks no problem for the name resolution, it is weird.  
And as 'xcatconfig -u' will be only run when `xCAT` package is upgraded, but it cannot give more information there because it does not run with `-V` option.

Not sure if it is possible for you to run 'xcatconfig -u -V' with `root` on you active MN only, and to see if the site.master will be changed.
If "Updated cluster site definition." is found in the output, it means the site table is updated.

Be careful, as doing this just like you upgrade xCAT via yum on the MN. Make sure you backup your data.

Here is output I'm running from an internal environment (2.13.9 Nov 21 build)
```
# xcatconfig -u -V
Running /opt/xcat/sbin/xcatconfig -u -V
Running command on briggs01: /bin/cat /etc/exports | grep '/tftpboot' 2>&1

Running command on briggs01: /bin/cat /etc/exports | grep '/install' 2>&1

Running command on briggs01: /bin/rm -rf /install/postscripts/ca >/dev/null 2>&1

Running command on briggs01: /bin/mkdir -m 755 -p /install/postscripts/ca 2>&1

Running command on briggs01: /bin/cp -p /etc/xcat/ca/ca-cert.pem /install/postscripts/ca/ca-cert.pem 2>&1

/bin/cp -p /etc/xcat/ca/ca-cert.pem /install/postscripts/ca/ca-cert.pem succeeded.
Running command on briggs01: /bin/rm -rf /install/postscripts/cert >/dev/null 2>&1

Running command on briggs01: /bin/rm -rf /install/postscripts/_xcat >/dev/null 2>&1

Running command on briggs01: /bin/mkdir -m 755 -p /install/postscripts/_xcat 2>&1

Created /install/postscripts/_xcat directory.
Running command on briggs01: /bin/cp -p /root/.xcat/ca.pem /install/postscripts/_xcat/ca.pem 2>&1

Copied /root/.xcat/ca.pem to /install/postscripts/_xcat directory.
Running command on briggs01: ip -4 --oneline addr show |awk -F ' ' '{print $4}'|awk -F '/' '{print $1}' 2>&1

Running command on briggs01: /bin/grep ^[^#]*nameserver /etc/resolv.conf | awk '{print $2}' 2>&1

Running command on briggs01: grep Subject /etc/xcat/cert/server-cert.pem 2>&1

Running command on briggs01: find /usr/share/zoneinfo -xtype f -exec cmp -s /etc/localtime {} \; -print | grep -v posix | grep -v SystemV | grep -v right | grep -v localtime  2>&1

Running command on briggs01: /opt/xcat/sbin/chtab priority=1.2 policy.name=briggs01.pok.stglabs.ibm.com policy.rule=trusted;/opt/xcat/sbin/chtab priority=2.3 policy.commands=lsxcatd policy.rule=allow;/opt/xcat/sbin/chtab priority=2.1 policy.commands=remoteimmsetup policy.rule=allow;/opt/xcat/sbin/chtab priority=4.9 policy.commands=getadapter policy.rule=allow; 2>&1

Updated policy definition for MN.
Running command on briggs01: /opt/xcat/sbin/chtab priority=4.7 policy.commands=litefile policy.rule=allow;/opt/xcat/sbin/chtab priority=4.8 policy.commands=litetree policy.rule=allow; 2>&1

Created statelite policy definition.
Running command on briggs01: /opt/xcat/sbin/chtab node=xcatdefaults postscripts.postbootscripts+='otherpkgs'; 2>&1

Updated postscripts definition.
Running command on briggs01: /opt/xcat/sbin/chtab -d node=xcatdefaults nodelist 2>&1

Removed xcatdefaults.
Running command on briggs01: XCATBYPASS=Y /opt/xcat/sbin/tabdump site 2>/dev/null |grep dnshandler 2>&1

Running command on briggs01: XCATBYPASS=Y /opt/xcat/sbin/tabdump site 2>/dev/null |grep xcatsslversion 2>&1

Running command on briggs01: /install/postscripts/syslog 2>&1

syslog has been set up.
Running command on briggs01: systemctl set-environment TERM=xterm-256color 2>&1

Imported TERM=xterm-256color into systemd.
Running command on briggs01: type systemctl >/dev/null 2>&1 && systemctl daemon-reload 2>&1

Restarting xcatd (via systemctl):                          [  OK  ]
Running command on briggs01: /usr/sbin/conserver -V 2>&1

Running command on briggs01: /bin/cat /etc/conserver.cf | grep sslauthority 2>&1
```
Note: there are no `Updated cluster site definition`.

This is just to narrow down the issues,  if it is hard for you to run this command,  I can modify the 
spec to make it run `xcatconfig -u -V` and logged to file for better debugging if hit issues.

**Comment by mattaezell** [comment]
I just ran `xcatconfig -u -V` on the master, and the `site.master` did NOT change.  Must be something else causing the change.  Is there some way to trace when the database changes?

**Comment by robin2008** [comment]
@mattaezell 
Suspect that it might be caused when run `yum update xCAT` on standby MN.
I change the spec to log the `xcatconfig -u -V` output to /var/log/xcat/upgrade.log in PR https://github.com/xcat2/xcat-core/pull/4389

And did you see any mkdef or chdef, chtab for `site` master in /var/log/commands.log*?

If the `site.master` is changed via `xcatd` daemon, we can enable `site.dbtracelevel` to 4 to trace all db operation in `xcatd`. (including query too)

But if the modification is not got through `xcatd`, for example, "XCATBYPASS" mode, or SQL directly, xcatd cannot monitor those events.

But we could change `postgresql` configuration to enable it log corresponding operations: 
For example `log_statement = 'mod'`
```
grep ^log_* postgresql.conf
logging_collector = on			# Enable capturing of stderr and csvlog
log_filename = 'postgresql-%a.log'	# log file name pattern,
log_truncate_on_rotation = on		# If on, an existing log file with the
log_rotation_age = 1d			# Automatic rotation of logfiles will
log_rotation_size = 0			# Automatic rotation of logfiles will
log_statement = 'mod'			# none, ddl, mod, all
log_timezone = 'US/Eastern'
```
After you restart postgresql service, you could check the log details in PG_DATA/pg_log/
```
LOG:  execute dbdpg_p8434_4: INSERT INTO site ("value","comments","key","disable") VALUES ($1,$2,$3,$4)
DETAIL:  parameters: $1 = '1', $2 = NULL, $3 = 'runbootscripts', $4 = NULL
LOG:  execute dbdpg_p8434_6: INSERT INTO site ("value","comments","key","disable") VALUES ($1,$2,$3,$4)
DETAIL:  parameters: $1 = '0', $2 = NULL, $3 = 'svloglocal', $4 = NULL
LOG:  execute dbdpg_p8434_8: INSERT INTO site ("value","comments","key","disable") VALUES ($1,$2,$3,$4)
DETAIL:  parameters: $1 = '64', $2 = NULL, $3 = 'blademaxp', $4 = NULL
```

Or do it with trigger:
https://wiki.postgresql.org/wiki/Audit_trigger_91plus

**Comment by whowutwut** [comment]
@robin2008 If we set the debug to that high, does it start to have impact on the performance of the xCAT daemon?

**Comment by adambertsch** [comment]
When we upgraded our xCAT master on our CORAL system we didn't see this issue.  We are currently only running with one xCAT master node.  That may lend more weight to this being related to upgrade of the standby/backup master node.

**Comment by mattaezell** [comment]
Found it!  The root cause is when we generate a stateless image with `xCATsn` installed.  We generally do that every time we upgrade xCAT, since we need the service node xCAT version to match.

The `xCATsn` RPM scriptlet has:
```
if [ -f "/proc/cmdline" ]; then   #check to make sure this is not image install
 if [ ! -s /etc/xcat/cfgloc ]; then  # database is sqlite
   $RPM_INSTALL_PREFIX0/sbin/xcatconfig -d
 fi
fi
```
I'm not sure exactly what yum does with `--installroot`, but apparently it still sees /proc/cmdline in the chroot.  We need a better test there.

Here's the output from `genimage`
```
  Installing : 4:xCATsn-2.13.9-snap201711282156.ppc64le                   77/82 
/bin/grep: /etc/resolv.conf: No such file or directory
Created symlink /etc/systemd/system/multi-user.target.wants/httpd.service, pointing to /usr/lib/systemd/system/httpd.service.
Restarting xcatd xcatd failed to start: Unable to perform socket takeover from existing xCAT instance
Unable to log xcatd failed to start: Unable to perform socket takeover from existing xCAT instance to syslog because of no connection to syslog available
        - unix dgram connect: Connection refused
        - stream can't open /dev/log: No such device or address at /opt/xcat/lib/perl/xCAT/MsgUtils.pm line 479.
[FAILED]

Running in chroot, ignoring request.
xCATsn is now installed
Error: Permission denied for request
Running "mknb ppc64le", triggered by the installation/update of xCAT-genesis-scripts-x86_64 ...
Error: Permission denied for request
warning: the client certificates under /root/.xcat/ are not setup correctly, please run '/opt/xcat/share/xcat/scripts/setup-local-client.sh ' as 'root' to generate the client certificates; otherwise, the SSL connection between xcat client and xcatd will be setup without certificate verification and open to Man-In-The-Middle attacks.
warning: %post(xCATsn-4:2.13.9-snap201711282156.ppc64le) scriptlet failed, exit status 1
Non-fatal POSTIN scriptlet failure in rpm package 4:xCATsn-2.13.9-snap201711282156.ppc64le
```

**Comment by robin2008** [comment]
@mattaezell 
Thanks for information, we will look at it quickly. Actually we did not found it in our env when genimage for SN also. As  two checking are there /proc/cmdline and /etc/xcat/cfgloc.

**Comment by immarvin** [comment]
hi @mattaezell , strange that /proc/cmdline exists in the rootimg dir, would you please provide the content of it? so that we can find where it come from

**Comment by mattaezell** [comment]
@immarvin It doesn't exist.  I think maybe /proc is actually mounted there?

**Comment by mattaezell** [comment]
I just made a simple RPM with a `%post` section and tried to install it using both RPM and YUM.  It doesn't seem that /proc is automatically mounted.

```
  Installing : testpost-1-1.el7a.ppc64le                                    1/1
mount: failed to read mtab: No such file or directory
findmnt: can't read /proc/mounts: No such file or directory
```
Must be something in our `genimage` that's mounting /proc

**Comment by mattaezell** [comment]
Oh, `genimage` has `mount_chroot` that mounts `proc` and `dev`.

**Comment by immarvin** [comment]
@mattaezell @robin2008 , this is the problem, 
in `rh/genimage`, the `/proc`, `/sys` and `/dev` under rootimg dir  are mounted from the host
```
 377     mount_chroot($rootimg_dir);
```
it is there because many %post scripts contain conditional operations based on the directories , I think we need better test in
```
if [ -f "/proc/cmdline" ]; then   #check to make sure this is not image install
 if [ ! -s /etc/xcat/cfgloc ]; then  # database is sqlite
   $RPM_INSTALL_PREFIX0/sbin/xcatconfig -d
 fi
fi
```

**Comment by mattaezell** [comment]
If we can be sure that `stat` is installed, you can use a test like follows as the second test:
```
if [ "$(stat -c '%i %d' /)" == "$(stat -c '%i %d' /proc/1/root/.)" ]; then
    echo not chroot
fi
```
That checks that that inode and device of the current / matches the inode and device of pid 1's /

**Comment by immarvin** [comment]
some discussion https://unix.stackexchange.com/questions/14345/how-do-i-tell-im-running-in-a-chroot on this

**Comment by whowutwut** [comment]
Fixed Using #4463 , GitHub has been collapsing events so it's not as easy to see when we associate the PR.

**Comment by robin2008** [comment]
Verified with stateless, stateful SN on fresh installation and update, work as normal.

As our env does not re-create the issue,  so close it first.
